### ISSUE TITLE
Utf8JsonReader stops reading with comment and JsonCommentHandling.Allow

### ISSUE THREAD

**Issue by IonKiwi** [dotnet/corefx#40385]
```
string json = @"
{
	""Property1"": [
		42
  ], // property2
  ""Property2"": 42
}
";

var reader = new Utf8JsonReader(Encoding.UTF8.GetBytes(json), new JsonReaderOptions() { AllowTrailingCommas = true, CommentHandling = JsonCommentHandling.Allow });
while (reader.Read()) {
	Console.WriteLine(reader.TokenType);
}
```

**With JsonCommentHandling.Allow**
the reader seems to stop reading after reading the comment.
The above code prints:
StartObject
PropertyName
StartArray
Number
EndArray
Comment

After the comment, reader.Read() returns false

**With JsonCommentHandling.Skip** the above code prints
StartObject
PropertyName
StartArray
Number
EndArray
**PropertyName
Number
EndObject**

It continues after reading/skipping the comment.

Version: .NET Core 3.0 Preview 8

**Comment by am11** [comment]
With my local testing, it seems like this patch fixes the bug:

```diff
--- a/src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs
+++ b/src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs
@@ -1989,7 +1989,7 @@ private ConsumeTokenResult ConsumeNextTokenFromLastNonCommentToken()
                     goto RollBack;
                 }
             }
-            else if (_tokenType == JsonTokenType.StartObject)
+            else if (_tokenType == JsonTokenType.StartObject || _tokenType == JsonTokenType.EndArray)
             {
                 if (first == JsonConstants.CloseBrace)
                 {
```

**Comment by ahsonkhan** [comment]
@IonKiwi, out of curiosity, what motivated you to use `JsonCommentHandling.Allow` rather than Skip?

**Comment by am11** [comment]
Same problem is with object:

```json
{
  "Property1" : {"Property1.1": 42},
  // comment
  "Property2": 42
}
```

this fixes both cases:

```diff
--- a/src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs
+++ b/src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs
@@ -1989,7 +1989,7 @@ private ConsumeTokenResult ConsumeNextTokenFromLastNonCommentToken()
                     goto RollBack;
                 }
             }
-            else if (_tokenType == JsonTokenType.StartObject)
+            else if (_tokenType == JsonTokenType.StartObject || _tokenType == JsonTokenType.EndObject || _tokenType == JsonTokenType.EndArray)
             {
                 if (first == JsonConstants.CloseBrace)
                 {
```

**Comment by ahsonkhan** [comment]
> Same problem is with object

Can you verify with both comment styles (`/*...*/` and `//... \r\n`) and with all comment endings (`\r`, `\n`, `\r\n`)? I am currently testing myself, especially for the multi-segment code path.
https://github.com/dotnet/corefx/blob/e15ae48c8538eeea0a4c6c29c14d3cc4390c54ae/src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs#L1730

**Comment by ahsonkhan** [comment]
@am11, if you are interested in making a fix, here are some starting tests. Although, I am not sure if the fix should be applied at that line, or handled separately within another if-block in `ConsumeNextTokenFromLastNonCommentToken`.

```C#

        [Theory]
        [InlineData("{\"Property1\": [ 42], // comment\n\"Property2\": 42}")]
        [InlineData("{\"Property1\": [ 42], // comment\r\"Property2\": 42}")]
        [InlineData("{\"Property1\": [ 42], // comment\r\n\"Property2\": 42}")]
        [InlineData("{\"Property1\": [ 42], /*comment*/\"Property2\": 42}")]
        [InlineData("{\"Property1\": {\"Property1.1\": 42}, // comment\n\"Property2\": 42}")]
        [InlineData("{\"Property1\": {\"Property1.1\": 42}, // comment\r\"Property2\": 42}")]
        [InlineData("{\"Property1\": {\"Property1.1\": 42}, // comment\r\n\"Property2\": 42}")]
        [InlineData("{\"Property1\": {\"Property1.1\": 42}, /*comment*/\"Property2\": 42}")]
        public static void ReadJsonCustom(string jsonString)
        {
            byte[] input = Encoding.UTF8.GetBytes(jsonString);
            var reader = new Utf8JsonReader(input, new JsonReaderOptions() { AllowTrailingCommas = true, CommentHandling = JsonCommentHandling.Allow });
            while (reader.Read())
            {
                Console.WriteLine(reader.TokenType);
            }
            Assert.Equal(input.Length, reader.BytesConsumed);
        }
```

**Comment by am11** [comment]
@ahsonkhan, i just get the new dev environment setup, was having some technical issues with the old one (related to native dependencies).. 😄 I can see that you already have the fix issued, so kudos. 👍 Meanwhile I found two failing tests in my machine, PR is up:  https://github.com/dotnet/corefx/pull/40440.

**Comment by ahsonkhan** [comment]
Re-opening until fixed in 3.0.

**Comment by ahsonkhan** [comment]
Fixed in release/3.0 in https://github.com/dotnet/corefx/pull/40443
